src/contexts/cardContext.tsx

```diff
@@ -0,0 +1,88 @@
+/* eslint-disable consistent-return */
+import { useReducer, createContext, Dispatch, ReactNode, useContext, useState, SetStateAction } from 'react'
+
+export interface State {
+  nickName: string
+  cardDesign: string
+  cardNumber: { num1: string; num2: string; num3: string; num4: string }
+  cardExpirationDate: { MM: string; YY: string }
+  ownerName: string
+  cardSecurityCode: string
+  cardPassword: { password1: string; password2: string }
+}
+
+type Action =
+  | {
+      type: 'CREATE'
+      card: State
+    }
+  | {
+      type: 'EDIT'
+      card: State
+    }
+  | {
+      type: 'REMOVE'
+      card: State
+    }
+
+const INIT_CARD = {
+  nickName: '',
+  cardDesign: '',
+  cardNumber: { num1: '', num2: '', num3: '', num4: '' },
+  cardExpirationDate: { MM: '', YY: '' },
+  ownerName: '',
+  cardSecurityCode: '',
+  cardPassword: { password1: '', password2: '' },
+}
+
+const cardReducer = (state: State[], action: Action) => {
+  switch (action.type) {
+    case 'CREATE':
+      return state.concat(action.card)
+    case 'EDIT':
+      return state.map((card: State) =>
+        card.ownerName === action.card.ownerName ? { ...card, nickName: action.card.nickName } : card
+      )
+    case 'REMOVE':
+      return state.filter((card: State) => card.ownerName !== action.card.ownerName)
+  }
+}
+
+const CardListStateContext = createContext<State[]>([])
+const CardDispatchContext = createContext<Dispatch<Action>>(() => null)
+const CardItemStateContext = createContext<State>(INIT_CARD)
+const CardItemSetContext = createContext<Dispatch<SetStateAction<State>>>(() => null)
+
+export const useCardListStateContext = () => {
+  const state = useContext(CardListStateContext)
+  return state
+}
+
+export const useCardDispatchContext = () => {
+  const dispatch = useContext(CardDispatchContext)
+  return dispatch
+}
+
+export const useCardItemStateContext = () => {
+  const itemState = useContext(CardItemStateContext)
+  return itemState
+}
+
+export const useCardItemSetContext = () => {
+  const setState = useContext(CardItemSetContext)
+  return setState
+}
```

Comment:
지금은 규모가 작고, 하나의 Context만 사용하고 있어서 이러한 context를 사용하는 hook을 잘못 사용할 일이 없겠지만 보다 큰 규모의 어플리케이션에서는 `useContext(CardItemSetContext)`가 실패 했을 경우도 생각해보시면 좋을거 같아요

Reply:
불필요한 할당 없이 이렇게 사용하시는건 어떨까요?

```suggestion
export const useCardItemSetContext = () => {
  return useContext(CardItemSetContext)
}
```

Reply:
좀 더 명시적으로 사용하려고 한 건데 다른 것들도 할당을 없애야 할까요?

Reply:
useCardItemSetContext로 예를 들면 useCardItemSetContextr가 반환하는건 State이지 `setState`라는 식별자가 아닌데요. 

실제로 사용할 때도 아래처럼 사용할 수도 있구요.

```
const asd = useCardItemSetContext()
```

개인 취향으로는 바로 return을 해주는걸 좋아하지만 수인님이 기존에 작성하신 방법을 선호하신다면 그 방법 그대로 사용하셔도 문제는 없어보이네요!